### PR TITLE
api: fix TLS when redirecting the request

### PR DESCRIPTION
**Pull request by rleungx**
### What problem does this PR solve? 
Start 2 PD: 
pd1: client-urls: 127.0.0.1:2379, peer-urls: 127.0.0.1:2380
pd2: client-urls: 127.0.0.1:2479, peer-urls: 127.0.0.1:2480

From PD log:
Before:
pd1:
```
[2019/09/09 13:54:31.332 +08:00] [ERROR] [redirector.go:97] ["request failed"] [error="Get https://127.0.0.1:2379/pd/ping: x509: certificate signed by unknown authority"]
[2019/09/09 13:55:31.332 +08:00] [ERROR] [redirector.go:97] ["request failed"] [error="Get https://127.0.0.1:2379/pd/ping: x509: certificate signed by unknown authority"]
```
pd2:
```
[2019/09/09 14:30:00.098 +08:00] [WARN] [config_logging.go:258] ["rejected connection"] [remote-addr=127.0.0.1:53098] [server-name=] [error="remote error: tls: bad certificate"]
[2019/09/09 14:30:01.366 +08:00] [WARN] [config_logging.go:258] ["rejected connection"] [remote-addr=127.0.0.1:53106] [server-name=] [error="remote error: tls: bad certificate"]
```
After: no such error

From curl:
Before:
```
curl --cacert /path/to/ca.pem --cert /path/to/pd-server.pem --key /path/to/pd-server-key.pem https://127.0.0.1:2479/pd/ping
redirect failed
```
After: no such error

### What is changed and how it works?
This PR fixes the TLS problem when redirecting the request.

### Check List 

Tests 

 - Manual test 

Related changes

 - Need to cherry-pick to the release branch
 - Need to be included in the release notes

### CODE REVIEW

File: server/api/redirector.go [hunk 1]
```diff
@@ -68,16 +69,28 @@ func (h *redirector) ServeHTTP(w http.ResponseWriter, r *http.Request, next http
...
 }
 
 type customReverseProxies struct {
 	urls   []url.URL
 	client *http.Client
 }
 
-func newCustomReverseProxies(urls []url.URL) *customReverseProxies {
+func newCustomReverseProxies(urls []url.URL, cfg *tls.Config) *customReverseProxies {
+	if cfg != nil {
```
Comment (Connor1996) [inline server/api/redirector.go]: seems no need

File: server/api/redirector.go [hunk 2]
```diff
@@ -68,16 +69,27 @@ func (h *redirector) ServeHTTP(w http.ResponseWriter, r *http.Request, next http
...
 }
 
 type customReverseProxies struct {
 	urls   []url.URL
 	client *http.Client
 }
 
-func newCustomReverseProxies(urls []url.URL) *customReverseProxies {
+func newCustomReverseProxies(urls []url.URL, cfg *tls.Config) *customReverseProxies {
+	dialClient = &http.Client{
```
Comment (disksing) [inline server/api/redirector.go]: Please don't recreate `dialClient` for every request

File: server/api/redirector.go [hunk 3]
```diff
@@ -68,16 +69,27 @@ func (h *redirector) ServeHTTP(w http.ResponseWriter, r *http.Request, next http
...
+		Transport: &http.Transport{
+			TLSClientConfig:   cfg,
+			DisableKeepAlives: true,
```
Comment (disksing) [inline server/api/redirector.go]: Can we enable keep alive? See https://github.com/pingcap/pd/issues/1718

Comment (rleungx) [reply server/api/redirector.go]: Maybe we can enable it in another PR.

**Comment by rleungx** [comment]
@onitake This fix will land in 3.0.5 soon.

**Comment by onitake** [comment]
@rleungx Thanks!

**Pull request merged**